python/gps/algorithm/algorithm_traj_opt_pilqr.py

```diff
@@ -0,0 +1,166 @@
+""" This file defines the PILQR-based algorithm. """
+import copy
+import logging
+import numpy as np
+
+from gps.algorithm.algorithm import Algorithm
+from gps.algorithm.config import ALG_PILQR
+
+from gps.algorithm.traj_opt.traj_opt_utils import approximated_cost
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class AlgorithmTrajOptPILQR(Algorithm):
+    def __init__(self, hyperparams):
+        config = copy.deepcopy(ALG_PILQR)
+        config.update(hyperparams)
+        Algorithm.__init__(self, config)
+
+    def iteration(self, sample_lists):
+        """
+        Run iteration of PILQR.
+        Args:
+            sample_lists: List of SampleList objects for each condition.
+        """
+        for m in range(self.M):
+            self.cur[m].sample_list = sample_lists[m]
+
+        # Update dynamics model using all samples.
+        self._update_dynamics()
+
+        self._update_step_size()  # KL Divergence step size.
+
+        # Run inner loop to compute new policies.
+        for _ in range(self._hyperparams['inner_iterations']):
+            self._update_trajectories()
+            self._advance_iteration_variables()
+
+    def _update_step_size(self):
+        """ Evaluate costs on samples, and adjust the step size. """
+        # Evaluate cost function for all conditions and samples.
+        for m in range(self.M):
+            self._eval_cost(m)
+
+        # Adjust step size relative to the previous iteration.
+        for m in range(self.M):
+            if self.iteration_count >= 1 and self.prev[m].sample_list:
+                self._stepadjust(m)
+
+    def _compute_step_pi2_lqr(self, m, pdpc, pdcc, cdcc, pmc, cmc, cur_res):
+        """
+        Calculate new step size for pi2 and lqr.
+        Args:
+        m: condition
+        pdpc: previous dynamics-previous controller cost
+        pdcc: previous dynamics-current controller cost
+        cdcc: current dynamics-current controller cost
+        pmc: previous actual cost
+        cmc: current actual cost
+        cur_res:
+        """
+
+    def _stepadjust(self, m):
+        """
+        Calculate new step sizes.
+        Args:
+            m: Condition
+        """
+        pdpc = self.traj_opt.estimate_cost(
```

Comment:
For readability, maybe add comment like "# previous dynamics, previous controller" here